### ISSUE TITLE
nonzero gradients for opts

### ISSUE THREAD

**Issue by loriab** [psi4/psi4#310]
@psi-rking, why is my gradient non-zero at the end of a successful geometry optimization, please? Below is runable input file with details of problem.

```
# these are the conditions, a purely conventional (we're going to be
#   comparing to Q-Chem) MP2/cc-pVTZ optimization

set basis cc-pvtz
set g_convergence gau_verytight
set scf_type pk
set mp2_type conv
set guess sad
set freeze_core false

# the gradient that any fully optimized molecule should match, more or less

zerograd = psi4.Matrix(5,3)
refgrad = [
    [     0.0, 0.0, 0.0],
    [     0.0, 0.0, 0.0],
    [     0.0, 0.0, 0.0],
    [     0.0, 0.0, 0.0],
    [     0.0, 0.0, 0.0]]
zerograd.set(refgrad)

# this is the first system, a fluoromethane that looks pretty good to the eye
#   zmat version, normz

molecule normz {
   C
   H 1 1.2
   H 1 1.2 2 104
   H 1 1.2 2 105 3 120
   F 1 1.3 2 106 4 121
}

# reference start NRE, E, G

norm_nre_start = 37.6824945882  # Q-Chem
norm_e_start = -139.5307182946  # Q-Chem
norm_g_start = psi4.Matrix(5, 3)  # Q-Chem
contents = [  # have to swap x & y, and -x
    [   -0.07990533,   -0.00642920,   -0.00348020,    0.04874047,    0.04107426],
    [   -0.04900557,   -0.00240633,    0.05677652,   -0.03006882,    0.02470420],
    [   -0.06715450,    0.05113530,   -0.00508959,   -0.00643364,    0.02754243]]
norm_g_start.set(zip(*contents))
norm_rmsg_start = 0.041309  # Q-Chem
# Full Analytical Gradient of MP2 Energy (in au.)
#             1             2             3             4             5
#    1   -0.04900557   -0.00240633    0.05677652   -0.03006882    0.02470420
#    2    0.07990533    0.00642920    0.00348020   -0.04874047   -0.04107426
#    3   -0.06715450    0.05113530   -0.00508959   -0.00643364    0.02754243

# reference optimized NRE, E (G should be 0)

nre_opt = 37.5276022922  # Q-Chem
e_opt = -139.5573922  # Q-Chem
#norm_g_opt  # Q-Chem
# Full Analytical Gradient of MP2 Energy (in au.)
#             1             2             3             4             5
#    1    0.00000025   -0.00000001   -0.00000021   -0.00000007    0.00000005
#    2    0.00000010    0.00000008    0.00000000    0.00000001   -0.00000020
#    3    0.00000038   -0.00000005   -0.00000006   -0.00000023   -0.00000003


def test_norm_start(mol):
    g0 = gradient('mp2', molecule=mol)
    compare_values(norm_nre_start, mol.nuclear_repulsion_energy(), 3, 'start nre')
    compare_values(norm_e_start, get_variable('CURRENT ENERGY'), 5, 'start ene')
    compare_values(norm_rmsg_start, g0.rms(), 4, 'start rms grad')
    compare_matrices(norm_g_start, g0, 3, 'start grad')

def test_zero_grad(mol):
    e, wfn = optimize('mp2', return_wfn=True, molecule=mol)
    g = wfn.gradient()
    g.print_out()
    compare_values(nre_opt, mol.nuclear_repulsion_energy(), 3, 'opt nre')
    compare_values(e_opt, get_variable('CURRENT ENERGY'), 5, 'opt ene')
    compare_matrices(zerograd, g, 3, 'opt ZERO grad')

def test_odd_start(mol):
    g0 = gradient('mp2', molecule=mol)
    compare_values(odd_nre_start, mol.nuclear_repulsion_energy(), 1, 'start nre')
    compare_values(odd_e_start, get_variable('CURRENT ENERGY'), 3, 'start ene')
    compare_values(odd_rmsg_start, g0.rms(), 4, 'start rms grad')
    compare_matrices(odd_g_start, g0, 3, 'start grad')

def test_badodd_grad(mol):
    e, wfn = optimize('mp2', return_wfn=True, molecule=mol)
    g = wfn.gradient()
    g.print_out()
    compare_values(odd_nre_opt, mol.nuclear_repulsion_energy(), 2, 'opt nre nogood')
    compare_values(odd_e_opt, get_variable('CURRENT ENERGY'), 5, 'opt ene nogood')
    compare_values(odd_rmsg_opt, g.rms(), 4, 'opt grad nogood')

test_norm_start(normz)
test_zero_grad(normz)

# And behold, the initial gradient matches Q-Chem, the optimization
#   progresses nicely, the RMS Force column goes to zero, and the gradient
#   at the end is zero.

#    start nre.........................................................PASSED
#    start ene.........................................................PASSED
#    start rms grad....................................................PASSED
#    start grad........................................................PASSED
#Optimizer: Optimization complete!
#    opt nre...........................................................PASSED
#    opt ene...........................................................PASSED
#    opt ZERO grad.....................................................PASSED

#  --------------------------------------------------------------------------------------------------------------- ~
#   Step         Total Energy             Delta E       MAX Force       RMS Force        MAX Disp        RMS Disp  ~
#  --------------------------------------------------------------------------------------------------------------- ~
#      1    -139.530718293045   -139.530718293045      0.05695995      0.03625240      0.19767161      0.12159384  ~
#      2    -139.555614729326     -0.024896436281      0.02341593      0.00968824      0.06857406      0.03977721  ~
#      3    -139.556998324542     -0.001383595215      0.00950087      0.00516674      0.02878292      0.01662559  ~
#      4    -139.557349947407     -0.000351622866      0.00371412      0.00163364      0.01005622      0.00527025  ~
#      5    -139.557391640737     -0.000041693330      0.00051457      0.00018711      0.00138018      0.00067210  ~
#      6    -139.557392223903     -0.000000583166      0.00004140      0.00002031      0.00028542      0.00012144  ~
#      7    -139.557392234146     -0.000000010243      0.00001186      0.00000566      0.00004411      0.00002377  ~
#      8    -139.557392234594     -0.000000000448      0.00000420      0.00000194      0.00001643      0.00000834  ~
#      9    -139.557392234645     -0.000000000051      0.00000176      0.00000074      0.00000686      0.00000306  ~
#     10    -139.557392234661     -0.000000000016      0.00000071      0.00000029      0.00000278      0.00000118  ~
#  --------------------------------------------------------------------------------------------------------------- ~


# Second system, a not-right by eye but still plausible geometry, oddz

molecule oddz {
   C
   H 1 1.792
   H 1 1.792 2  59.4
   H 1 1.089 2 136.7 3 -127.3
   F 1 2.089 2 101.7 4 -135.9
}

# reference start NRE, E, G for this geometry

odd_nre_start = 25.7790014493  # Q-Chem
odd_e_start = -139.1749807824  # Q-Chem
# Full Analytical Gradient of MP2 Energy (in au.)
#             1             2             3             4             5
#    1   -0.06704124    0.03019136    0.05220334   -0.00529778   -0.01005568
#    2    0.09718540   -0.00309814   -0.00309734   -0.00009178   -0.09089814
#    3   -0.11760428    0.07849561    0.06595344   -0.00926729   -0.01757747
odd_g_start = psi4.Matrix(5, 3)  # Q-Chem
contents = [  # have to swap x & y, and -x
    [   -0.09718540,    0.00309814,    0.00309734,    0.00009178,    0.09089814],
    [   -0.06704124,    0.03019136,    0.05220334,   -0.00529778,   -0.01005568],
    [   -0.11760428,    0.07849561,    0.06595344,   -0.00926729,   -0.01757747]]
odd_g_start.set(zip(*contents))
odd_rmsg_start = 0.058156

# optimized reference should be same as norm, but in practice is this

odd_nre_opt = 33.838291620
odd_e_opt = -139.301327453
odd_rmsg_opt = 0.0422103431176

# should be able to pass starting gradient at new geometry, then zero grad final structure.

test_odd_start(oddz)  # works
#test_zero_grad(oddz)  # should work
test_badodd_grad(oddz)  # works b/c I'm testing against the bad results

# qu-est-ce que c-est? By RMS Force the optimization looks to proceed 
#   nicely. But look at that final gradient with unmoveable large gradient
#   components. It's not orientation, as input is zmat and it's the same with cartesian 
#   coordinates in a couple orientations. It's not corruption associated
#   with wavefunction pass, as I've tested an older version. It's not a local 
#   minimum as the total gradient wouldn't be so large. It's not too distorted a 
#   starting structure, as q-chem and molpro converge to the real minimum promptly. It's not
#   symmetry constrained, as it's c1. I'm at a loss.

# for oddz
#   start
#   25.7790013416
#   -139.174980786
#   0.0581560524872
#   Optimizer: Optimization complete!
#   opt
#   33.8382916197
#   -139.301327453
#   0.0422103431176
#gradient = [
#    [    -0.01135649578523,    -0.02871892445843,    -0.05475995417434],
#    [     0.00216387360216,     0.04113784938512,     0.07821983299617],
#    [     0.00215642675781,     0.04114952366672,     0.07824201651882],
#    [     0.00563414370780,    -0.03413386327141,    -0.06481576207381],
#    [     0.00140205171291,    -0.01943458532669,    -0.03688613326736]]

#  --------------------------------------------------------------------------------------------------------------- ~
#   Step         Total Energy             Delta E       MAX Force       RMS Force        MAX Disp        RMS Disp  ~
#  --------------------------------------------------------------------------------------------------------------- ~
#      1    -139.174980785934   -139.174980785934      0.08977532      0.06906505      0.39164409      0.25000003  ~
#      2    -139.237253751032     -0.062272965098      0.08963832      0.05551392      0.44053120      0.25000000  ~
#      3    -139.282037341604     -0.044783590572      0.06072927      0.03180855      0.57480436      0.30215430  ~
#      4    -139.291415380240     -0.009378038636      0.07020602      0.04747338      0.40870734      0.23074492  ~
#      5    -139.289103767764      0.002311612476      0.06272061      0.03636254      0.31265835      0.16807142  ~
#      6    -139.299937489944     -0.010833722180      0.02206926      0.01444069      0.08604223      0.05017704  ~
#      7    -139.301130513523     -0.001193023579      0.00601199      0.00414213      0.03851787      0.02297290  ~
#      8    -139.301272108227     -0.000141594704      0.00392137      0.00273882      0.01100483      0.00820114  ~
#      9    -139.301322123849     -0.000050015621      0.00099031      0.00070170      0.00667642      0.00454418  ~
#     10    -139.301327446273     -0.000005322424      0.00004217      0.00002303      0.00027466      0.00017580  ~
#     11    -139.301327453303     -0.000000007030      0.00000289      0.00000156      0.00001815      0.00001211  ~
#     12    -139.301327453322     -0.000000000020      0.00000018      0.00000009      0.00000104      0.00000060  ~
#  --------------------------------------------------------------------------------------------------------------- ~
```

**Comment by psi-rking** [comment]
The optimizer is not creating covalent bond coordinates for the C-H
distances which in the latter geometry are 1.8 Angstroms.  The optimizer
tried to loosen the criterian for bonding until everything was bonded to
something, but not successfully coordinating the full system.  So the
system is underdetermined in internal coordinates (take a look at the
coordinate list in the output).

If one wants to add bonds for unusual distances you can do it with, in this
case,

covalent_connect 1.8  # default is 1.3

For more detailed control, the intco file can be modified.

**Comment by loriab** [comment]
Shouldn't the mismatch between the total gradient rms (0.04) and the gradient it's decided to consider (1e-7) signal to the optimizer that it doesn't have the full picture? Throw in some Cartesian steps or whatever others do?

**Comment by psi-rking** [comment]
Agreed, this should be improved.  However, there are various reasons why
someone might want to optimize only along a restricted set of coordinates,
so I am not sure this should go automatically to error mode (switching to
cartesians, etc.).  Certainly, I should add a WARNING at the end that the
full space wasn't optimized.

**Comment by loriab** [comment]
Certainly the flexibility of the optimizer oughtn't be reduced. In the absence of fixed/frozen keywords or user supplied/edited intco (do you need a keyword to signal do-read-intco instead of accepting any it finds?), it's probably reasonable to assume a full optimization is desired. Maybe ``grep '~'`` needs to catch a `gradient.rms()`, too, next to the optking-active rms gradient.

**Comment by psi-rking** [comment]
The current code detected 4 different fragments for the extended CH3F.
Then it tried to connect them.  It really should have been able to do so.
I don't see why it didn't work, and I'm looking for a bug.

**Comment by psi-rking** [comment]
I have fixed the code that connects the initial, separated fragments in
optking.  The CH3F case now runs fine.  The code was very much in error.  I
will submit the correction whenever the main psi4 trunk runs again.  At the
moment, the python there seems broken.

**Comment by loriab** [comment]
The current head (0.4.156 / 7783818 / 0.5.92) works fine for me on the test case above and on all the usual test cases. What is the nature of your broken python?

**Comment by psi-rking** [comment]
An error has occurred. Traceback:

<class 'Boost.Python.ArgumentError'>: Python argument types in

    psi4.scf(NoneType, NoneType)

did not match C++ signature:

    scf(boost::shared_ptr<psi::Wavefunction>, _object*, _object*):   File
"<string>", line 47, in <module>

This is, for example, from the opt1 test case, but all jobs fail.  I
already tried removing my objdir completely and recompiling.

**Comment by psi-rking** [comment]
Have we changed what PSIDATADIR should point to?

**Comment by psi-rking** [comment]
With

setenv PSIDATADIR /home/rking/psi4public/share

An error has occurred. Traceback:

<type 'exceptions.SyntaxError'>: ('invalid syntax',
('/home/rking/psi4public/share/python/qcdb/dbwrap.py', 1455, 47, '
return {db: odb.integer_reactions() for db, odb in
self.dbdict.items()}\n')):   File "<string>", line 7, in <module>

**Comment by loriab** [comment]
Oh bother, python 2.6 doesn't have dictionary comprehensions. I updated qcdb within the last couple weeks with accumulated improvements and python 3.5 prep.

It doesn't look like I use dictionary comprehensions all that much, so you can try the following, then report any subsequent errors.
```
        return {db: odb.integer_reactions() for db, odb in self.dbdict.items()}  # replace this
        return dict((db, odb.integer_reactions()) for db, odb in self.dbdict.items())  # with this
```

Or, would you like to give python 2.7 a try? Do the binary instructions (http://psicode.org/psi4manual/master/conda.html#quick-installation) through the `conda update` step (replace "Linux" with "MacOSX" in installer file name, if on Mac), be sure your `which python` points to 2.7, then compile psi4 with `--python $MINICONDA/bin/python`. When we get around to supporting python3.5, python2.6 support may go.

**Comment by psi-rking** [comment]
OK, switched to python 2.7 and all is fine.
-Rollin

**Comment by loriab** [comment]
Thanks! fyi, if you include "fixes #310" in the commit message, it'll close the issue when the PR gets merged.
